File: packages/less/src/less/tree/value.test.js
import { describe, it, expect, vi } from 'vitest';
import Value from './value';
import Node from './node';

describe('Value', () => {
    describe('constructor', () => {
        it('should throw error when constructed without an argument', () => {
            expect(() => new Value()).toThrow(
                'Value requires an array argument'
            );
        });

        it('should throw error when constructed with null', () => {
            expect(() => new Value(null)).toThrow(
                'Value requires an array argument'
            );
        });

        it('should throw error when constructed with undefined', () => {
            expect(() => new Value(undefined)).toThrow(
                'Value requires an array argument'
            );
        });

        it('should wrap single non-array value in array', () => {
            const singleValue = { type: 'test' };
            const value = new Value(singleValue);
            expect(value.value).toEqual([singleValue]);
        });

        it('should store array value as is', () => {
            const arrayValue = [{ type: 'test1' }, { type: 'test2' }];
            const value = new Value(arrayValue);
            expect(value.value).toBe(arrayValue);
        });

        it('should handle empty array', () => {
            const value = new Value([]);
            expect(value.value).toEqual([]);
        });

        it('should inherit from Node', () => {
            const value = new Value([]);
            expect(value instanceof Node).toBe(true);
            expect(value.type).toBe('Value');
        });

        it('should handle array with null/undefined values', () => {
            const arrayValue = [null, undefined, { type: 'test' }];
            const value = new Value(arrayValue);
            expect(value.value).toBe(arrayValue);
        });

        it('should handle array with primitive values', () => {
            const arrayValue = [123, 'string', true];
            const value = new Value(arrayValue);
            expect(value.value).toBe(arrayValue);
        });
    });

    describe('accept', () => {
        it('should call visitArray on visitor with value array', () => {
            const mockVisitor = {
                visitArray: vi.fn((arr) => arr)
            };
            const valueArr = [{ type: 'test' }];
            const value = new Value(valueArr);

            value.accept(mockVisitor);
            expect(mockVisitor.visitArray).toHaveBeenCalledWith(valueArr);
        });

        it('should not call visitArray if value is undefined', () => {
            const mockVisitor = {
                visitArray: vi.fn((arr) => arr)
            };
            const value = new Value([]);
            value.value = undefined;

            value.accept(mockVisitor);
            expect(mockVisitor.visitArray).not.toHaveBeenCalled();
        });

        it('should not call visitArray if value is null', () => {
            const mockVisitor = {
                visitArray: vi.fn((arr) => arr)
            };
            const value = new Value([]);
            value.value = null;

            value.accept(mockVisitor);
            expect(mockVisitor.visitArray).not.toHaveBeenCalled();
        });

        it('should assign visitArray result back to value', () => {
            const transformedValue = [{ type: 'transformed' }];
            const mockVisitor = {
                visitArray: vi.fn(() => transformedValue)
            };
            const value = new Value([{ type: 'original' }]);

            value.accept(mockVisitor);
            expect(value.value).toBe(transformedValue);
        });

        it('should handle visitor returning null', () => {
            const mockVisitor = {
                visitArray: vi.fn().mockReturnValueOnce(null)
            };
            const value = new Value([{ type: 'test' }]);

            value.accept(mockVisitor);
            expect(value.value).toBeNull();

            // Verify that null value prevents future visitArray calls
            value.accept(mockVisitor);
            expect(mockVisitor.visitArray).toHaveBeenCalledTimes(1);
        });

        it('should handle visitor returning undefined', () => {
            const mockVisitor = {
                visitArray: vi.fn().mockReturnValueOnce(undefined)
            };
            const value = new Value([{ type: 'test' }]);

            value.accept(mockVisitor);
            expect(value.value).toBeUndefined();

            // Verify that undefined value prevents future visitArray calls
            value.accept(mockVisitor);
            expect(mockVisitor.visitArray).toHaveBeenCalledTimes(1);
        });

        it('should handle visitor throwing an error', () => {
            const mockVisitor = {
                visitArray: vi.fn(() => {
                    throw new Error('Visitor error');
                })
            };
            const value = new Value([{ type: 'test' }]);

            expect(() => value.accept(mockVisitor)).toThrow('Visitor error');
        });
    });

    describe('eval', () => {
        it('should return evaluated single value', () => {
            const mockContext = { someContext: true };
            const mockEvalResult = { type: 'evaluated' };
            const mockValue = {
                eval: vi.fn(() => mockEvalResult)
            };

            const value = new Value([mockValue]);
            const result = value.eval(mockContext);

            expect(mockValue.eval).toHaveBeenCalledWith(mockContext);
            expect(result).toBe(mockEvalResult);
        });

        it('should handle empty array', () => {
            const value = new Value([]);
            const result = value.eval({});
            expect(result).toBeInstanceOf(Value);
            expect(result.value).toEqual([]);
        });

        it('should return new Value with evaluated multiple values', () => {
            const mockContext = { someContext: true };
            const mockValues = [
                { eval: vi.fn(() => ({ type: 'eval1' })) },
                { eval: vi.fn(() => ({ type: 'eval2' })) }
            ];

            const value = new Value(mockValues);
            const result = value.eval(mockContext);

            expect(result).toBeInstanceOf(Value);
            expect(result.value).toEqual([
                { type: 'eval1' },
                { type: 'eval2' }
            ]);
            mockValues.forEach((v) => {
                expect(v.eval).toHaveBeenCalledWith(mockContext);
            });
        });

        it('should evaluate values in order', () => {
            const evaluationOrder = [];
            const mockValues = [
                {
                    eval: vi.fn(() => {
                        evaluationOrder.push(1);
                        return { type: 'eval1' };
                    })
                },
                {
                    eval: vi.fn(() => {
                        evaluationOrder.push(2);
                        return { type: 'eval2' };
                    })
                },
                {
                    eval: vi.fn(() => {
                        evaluationOrder.push(3);
                        return { type: 'eval3' };
                    })
                }
            ];

            const value = new Value(mockValues);
            value.eval({});

            expect(evaluationOrder).toEqual([1, 2, 3]);
        });

        it('should handle values that throw during evaluation', () => {
            const mockValue = {
                eval: vi.fn(() => {
                    throw new Error('Eval error');
                })
            };
            const value = new Value([mockValue]);

            expect(() => value.eval({})).toThrow('Eval error');
        });

        it('should handle values returning null/undefined during evaluation', () => {
            const mockValues = [
                { eval: vi.fn(() => null) },
                { eval: vi.fn(() => undefined) }
            ];
            const value = new Value(mockValues);
            const result = value.eval({});

            expect(result).toBeInstanceOf(Value);
            expect(result.value).toEqual([null, undefined]);
        });

        it('should handle values returning primitive types during evaluation', () => {
            const mockValues = [
                { eval: vi.fn(() => 123) },
                { eval: vi.fn(() => 'string') },
                { eval: vi.fn(() => true) }
            ];
            const value = new Value(mockValues);
            const result = value.eval({});

            expect(result).toBeInstanceOf(Value);
            expect(result.value).toEqual([123, 'string', true]);
        });
    });

    describe('genCSS', () => {
        it('should handle empty array', () => {
            const mockOutput = { add: vi.fn() };
            const value = new Value([]);
            value.genCSS({}, mockOutput);
            expect(mockOutput.add).not.toHaveBeenCalled();
        });

        it('should generate CSS for single value', () => {
            const mockOutput = { add: vi.fn() };
            const mockValue = {
                genCSS: vi.fn((ctx, out) => out.add('test'))
            };

            const value = new Value([mockValue]);
            value.genCSS({}, mockOutput);

            expect(mockValue.genCSS).toHaveBeenCalledWith({}, mockOutput);
            expect(mockOutput.add).toHaveBeenCalledWith('test');
        });

        it('should generate CSS for multiple values with default separator', () => {
            const mockOutput = { add: vi.fn() };
            const mockValues = [
                { genCSS: vi.fn((ctx, out) => out.add('test1')) },
                { genCSS: vi.fn((ctx, out) => out.add('test2')) }
            ];

            const value = new Value(mockValues);
            value.genCSS({}, mockOutput);

            expect(mockOutput.add).toHaveBeenCalledWith('test1');
            expect(mockOutput.add).toHaveBeenCalledWith(', ');
            expect(mockOutput.add).toHaveBeenCalledWith('test2');
        });

        it('should generate CSS for multiple values with compressed separator', () => {
            const mockOutput = { add: vi.fn() };
            const mockValues = [
                { genCSS: vi.fn((ctx, out) => out.add('test1')) },
                { genCSS: vi.fn((ctx, out) => out.add('test2')) }
            ];

            const value = new Value(mockValues);
            value.genCSS({ compress: true }, mockOutput);

            expect(mockOutput.add).toHaveBeenCalledWith('test1');
            expect(mockOutput.add).toHaveBeenCalledWith(',');
            expect(mockOutput.add).toHaveBeenCalledWith('test2');
        });

        it('should not add separator after last value', () => {
            const mockOutput = { add: vi.fn() };
            const mockValues = [
                { genCSS: vi.fn((ctx, out) => out.add('test1')) },
                { genCSS: vi.fn((ctx, out) => out.add('test2')) }
            ];

            const value = new Value(mockValues);
            value.genCSS({}, mockOutput);

            const calls = mockOutput.add.mock.calls;
            expect(calls[calls.length - 1][0]).toBe('test2');
        });

        it('should pass context to each value genCSS call', () => {
            const mockContext = { compress: true, someOtherFlag: true };
            const mockValues = [{ genCSS: vi.fn() }, { genCSS: vi.fn() }];

            const value = new Value(mockValues);
            value.genCSS(mockContext, { add: vi.fn() });

            mockValues.forEach((v) => {
                expect(v.genCSS).toHaveBeenCalledWith(
                    mockContext,
                    expect.any(Object)
                );
            });
        });

        it('should generate CSS values in order', () => {
            const generationOrder = [];
            const mockOutput = { add: vi.fn() };
            const mockValues = [
                {
                    genCSS: vi.fn(() => {
                        generationOrder.push(1);
                        mockOutput.add('test1');
                    })
                },
                {
                    genCSS: vi.fn(() => {
                        generationOrder.push(2);
                        mockOutput.add('test2');
                    })
                },
                {
                    genCSS: vi.fn(() => {
                        generationOrder.push(3);
                        mockOutput.add('test3');
                    })
                }
            ];

            const value = new Value(mockValues);
            value.genCSS({}, mockOutput);

            expect(generationOrder).toEqual([1, 2, 3]);
        });

        it('should handle values throwing during CSS generation', () => {
            const mockValue = {
                genCSS: vi.fn(() => {
                    throw new Error('GenCSS error');
                })
            };
            const value = new Value([mockValue]);

            expect(() => value.genCSS({}, { add: vi.fn() })).toThrow(
                'GenCSS error'
            );
        });

        it('should handle values with empty/whitespace-only output', () => {
            const mockOutput = { add: vi.fn() };
            const mockValues = [
                { genCSS: vi.fn((ctx, out) => out.add('')) },
                { genCSS: vi.fn((ctx, out) => out.add('   ')) },
                { genCSS: vi.fn((ctx, out) => out.add('test')) }
            ];

            const value = new Value(mockValues);
            value.genCSS({}, mockOutput);

            // Should still add separators between values
            expect(mockOutput.add).toHaveBeenCalledWith('');
            expect(mockOutput.add).toHaveBeenCalledWith(', ');
            expect(mockOutput.add).toHaveBeenCalledWith('   ');
            expect(mockOutput.add).toHaveBeenCalledWith(', ');
            expect(mockOutput.add).toHaveBeenCalledWith('test');
        });

        it('should handle output.add throwing an error', () => {
            const mockOutput = {
                add: vi.fn(() => {
                    throw new Error('Output error');
                })
            };
            const value = new Value([
                { genCSS: vi.fn((ctx, out) => out.add('test')) }
            ]);

            expect(() => value.genCSS({}, mockOutput)).toThrow('Output error');
        });

        it('should handle various context compression settings', () => {
            const mockOutput = { add: vi.fn() };
            const mockValues = [
                { genCSS: vi.fn((ctx, out) => out.add('test1')) },
                { genCSS: vi.fn((ctx, out) => out.add('test2')) }
            ];
            const value = new Value(mockValues);

            // Test different compression settings
            value.genCSS({ compress: true }, mockOutput);
            expect(mockOutput.add).toHaveBeenCalledWith(',');

            mockOutput.add.mockClear();
            value.genCSS({ compress: false }, mockOutput);
            expect(mockOutput.add).toHaveBeenCalledWith(', ');

            mockOutput.add.mockClear();
            value.genCSS({ compress: undefined }, mockOutput);
            expect(mockOutput.add).toHaveBeenCalledWith(', ');
        });
    });
});

File: packages/less/src/less/go_parser/value_test.go
package go_parser

import (
	"reflect"
	"testing"
)

// valueTestVisitor implements the visitor interface for testing
type valueTestVisitor struct {
	visitArrayFn func([]any) []any
}

func (v *valueTestVisitor) VisitArray(arr []any) []any {
	if v.visitArrayFn != nil {
		return v.visitArrayFn(arr)
	}
	return arr
}

func TestValue(t *testing.T) {
	t.Run("constructor", func(t *testing.T) {
		t.Run("should throw error when constructed without an argument", func(t *testing.T) {
			_, err := NewValue(nil)
			if err == nil || err.Error() != "Value requires an array argument" {
				t.Error("Expected error when constructed without an argument")
			}
		})

		t.Run("should handle array with null/undefined values", func(t *testing.T) {
			arrayValue := []any{nil, nil, map[string]string{"type": "test"}}
			value, err := NewValue(arrayValue)
			if err != nil {
				t.Errorf("Unexpected error: %v", err)
			}
			if !reflect.DeepEqual(value.Value, arrayValue) {
				t.Error("Expected array with nil values to be preserved")
			}
		})

		t.Run("should handle array with primitive values", func(t *testing.T) {
			arrayValue := []any{123, "string", true}
			value, err := NewValue(arrayValue)
			if err != nil {
				t.Errorf("Unexpected error: %v", err)
			}
			if !reflect.DeepEqual(value.Value, arrayValue) {
				t.Error("Expected array with primitive values to be preserved")
			}
		})

		t.Run("should wrap single non-array value in array", func(t *testing.T) {
			singleValue := map[string]string{"type": "test"}
			value, err := NewValue(singleValue)
			if err != nil {
				t.Errorf("Unexpected error: %v", err)
			}
			if len(value.Value) != 1 || !reflect.DeepEqual(value.Value[0], singleValue) {
				t.Error("Expected single value to be wrapped in array")
			}
		})

		t.Run("should store array value as is", func(t *testing.T) {
			arrayValue := []any{
				map[string]string{"type": "test1"},
				map[string]string{"type": "test2"},
			}
			value, err := NewValue(arrayValue)
			if err != nil {
				t.Errorf("Unexpected error: %v", err)
			}
			if len(value.Value) != 2 {
				t.Error("Expected array length to be preserved")
			}
			if !reflect.DeepEqual(value.Value[0], map[string]string{"type": "test1"}) {
				t.Error("Expected first value to be preserved")
			}
			if !reflect.DeepEqual(value.Value[1], map[string]string{"type": "test2"}) {
				t.Error("Expected second value to be preserved")
			}
		})

		t.Run("should handle empty array", func(t *testing.T) {
			value, err := NewValue([]any{})
			if err != nil {
				t.Errorf("Unexpected error: %v", err)
			}
			if len(value.Value) != 0 {
				t.Error("Expected empty array to be preserved")
			}
		})

		t.Run("should inherit from Node", func(t *testing.T) {
			value, err := NewValue([]any{})
			if err != nil {
				t.Errorf("Unexpected error: %v", err)
			}
			if value.Node == nil {
				t.Error("Expected value to inherit from Node")
			}
			if value.GetType() != "Value" {
				t.Error("Expected type to be 'Value'")
			}
		})
	})

	t.Run("accept", func(t *testing.T) {
		t.Run("should call visitArray on visitor with value array", func(t *testing.T) {
			valueArr := []any{map[string]string{"type": "test"}}
			value, _ := NewValue(valueArr)
			called := false
			visitor := &valueTestVisitor{
				visitArrayFn: func(arr []any) []any {
					called = true
					if len(arr) != 1 || !reflect.DeepEqual(arr[0], map[string]string{"type": "test"}) {
						t.Error("Unexpected array passed to visitArray")
					}
					return arr
				},
			}
			value.Accept(visitor)
			if !called {
				t.Error("Expected visitArray to be called")
			}
		})

		t.Run("should not call visitArray if value is undefined", func(t *testing.T) {
			value, _ := NewValue([]any{})
			value.Value = nil
			called := false
			visitor := &valueTestVisitor{
				visitArrayFn: func(arr []any) []any {
					called = true
					return arr
				},
			}
			value.Accept(visitor)
			if called {
				t.Error("Expected visitArray not to be called")
			}
		})

		t.Run("should assign visitArray result back to value", func(t *testing.T) {
			transformedValue := []any{map[string]string{"type": "transformed"}}
			value, _ := NewValue([]any{map[string]string{"type": "original"}})
			visitor := &valueTestVisitor{
				visitArrayFn: func(arr []any) []any {
					return transformedValue
				},
			}
			value.Accept(visitor)
			if len(value.Value) != 1 || !reflect.DeepEqual(value.Value[0], map[string]string{"type": "transformed"}) {
				t.Error("Expected value to be transformed")
			}
		})
	})

	t.Run("eval", func(t *testing.T) {
		t.Run("should return evaluated single value", func(t *testing.T) {
			mockContext := map[string]any{"someContext": true}
			mockEvalResult := map[string]string{"type": "evaluated"}
			mockValue := &mockEvaluator{
				evalFn: func(ctx any) (any, error) {
					if !reflect.DeepEqual(ctx, mockContext) {
						t.Error("Expected context to be passed to eval")
					}
					return mockEvalResult, nil
				},
			}
			value, _ := NewValue([]any{mockValue})
			result, err := value.Eval(mockContext)
			if err != nil {
				t.Errorf("Unexpected error: %v", err)
			}
			if !reflect.DeepEqual(result, mockEvalResult) {
				t.Error("Expected result to be evaluated value")
			}
		})

		t.Run("should handle empty array", func(t *testing.T) {
			value, _ := NewValue([]any{})
			result, err := value.Eval(map[string]any{})
			if err != nil {
				t.Errorf("Unexpected error: %v", err)
			}
			resultValue, ok := result.(*Value)
			if !ok {
				t.Error("Expected result to be Value")
			}
			if len(resultValue.Value) != 0 {
				t.Error("Expected result to be empty array")
			}
		})

		t.Run("should return new Value with evaluated multiple values", func(t *testing.T) {
			mockContext := map[string]any{"someContext": true}
			mockValues := []any{
				&mockEvaluator{
					evalFn: func(ctx any) (any, error) {
						return map[string]string{"type": "eval1"}, nil
					},
				},
				&mockEvaluator{
					evalFn: func(ctx any) (any, error) {
						return map[string]string{"type": "eval2"}, nil
					},
				},
			}
			value, _ := NewValue(mockValues)
			result, err := value.Eval(mockContext)
			if err != nil {
				t.Errorf("Unexpected error: %v", err)
			}
			resultValue, ok := result.(*Value)
			if !ok {
				t.Error("Expected result to be Value")
			}
			if len(resultValue.Value) != 2 {
				t.Error("Expected result to have two values")
			}
			if !reflect.DeepEqual(resultValue.Value[0], map[string]string{"type": "eval1"}) {
				t.Error("Expected first value to be eval1")
			}
			if !reflect.DeepEqual(resultValue.Value[1], map[string]string{"type": "eval2"}) {
				t.Error("Expected second value to be eval2")
			}
		})

		t.Run("should evaluate values in order", func(t *testing.T) {
			evaluationOrder := []int{}
			mockValues := []any{
				&mockEvaluator{
					evalFn: func(ctx any) (any, error) {
						evaluationOrder = append(evaluationOrder, 1)
						return map[string]string{"type": "eval1"}, nil
					},
				},
				&mockEvaluator{
					evalFn: func(ctx any) (any, error) {
						evaluationOrder = append(evaluationOrder, 2)
						return map[string]string{"type": "eval2"}, nil
					},
				},
				&mockEvaluator{
					evalFn: func(ctx any) (any, error) {
						evaluationOrder = append(evaluationOrder, 3)
						return map[string]string{"type": "eval3"}, nil
					},
				},
			}

			value, _ := NewValue(mockValues)
			_, err := value.Eval(map[string]any{})
			if err != nil {
				t.Errorf("Unexpected error: %v", err)
			}

			if !reflect.DeepEqual(evaluationOrder, []int{1, 2, 3}) {
				t.Error("Expected values to be evaluated in order")
			}
		})

		t.Run("should handle values returning primitive types during evaluation", func(t *testing.T) {
			mockValues := []any{
				&mockEvaluator{
					evalFn: func(ctx any) (any, error) {
						return 123, nil
					},
				},
				&mockEvaluator{
					evalFn: func(ctx any) (any, error) {
						return "string", nil
					},
				},
				&mockEvaluator{
					evalFn: func(ctx any) (any, error) {
						return true, nil
					},
				},
			}
			value, _ := NewValue(mockValues)
			result, err := value.Eval(map[string]any{})
			if err != nil {
				t.Errorf("Unexpected error: %v", err)
			}

			resultValue, ok := result.(*Value)
			if !ok {
				t.Error("Expected result to be Value")
			}
			expectedValues := []any{123, "string", true}
			if !reflect.DeepEqual(resultValue.Value, expectedValues) {
				t.Error("Expected primitive values to be preserved")
			}
		})
	})

	t.Run("genCSS", func(t *testing.T) {
		t.Run("should handle empty array", func(t *testing.T) {
			value, _ := NewValue([]any{})
			var output []string
			outputFn := func(chunk any, fileInfo any, index any) {
				output = append(output, chunk.(string))
			}
			value.GenCSS(map[string]any{}, &CSSOutput{Add: outputFn})
			if len(output) != 0 {
				t.Error("Expected no output for empty array")
			}
		})

		t.Run("should generate CSS for single value", func(t *testing.T) {
			mockValue := &mockCSSGenerator{
				genCSSFn: func(ctx any, out *CSSOutput) {
					out.Add("test", nil, nil)
				},
			}
			value, _ := NewValue([]any{mockValue})
			var output []string
			outputFn := func(chunk any, fileInfo any, index any) {
				output = append(output, chunk.(string))
			}
			value.GenCSS(map[string]any{}, &CSSOutput{Add: outputFn})
			if len(output) != 1 || output[0] != "test" {
				t.Error("Expected single value output")
			}
		})

		t.Run("should generate CSS for multiple values with default separator", func(t *testing.T) {
			mockValues := []any{
				&mockCSSGenerator{
					genCSSFn: func(ctx any, out *CSSOutput) {
						out.Add("test1", nil, nil)
					},
				},
				&mockCSSGenerator{
					genCSSFn: func(ctx any, out *CSSOutput) {
						out.Add("test2", nil, nil)
					},
				},
			}
			value, _ := NewValue(mockValues)
			var output []string
			outputFn := func(chunk any, fileInfo any, index any) {
				output = append(output, chunk.(string))
			}
			value.GenCSS(map[string]any{}, &CSSOutput{Add: outputFn})
			if len(output) != 3 || output[0] != "test1" || output[1] != ", " || output[2] != "test2" {
				t.Error("Expected output with default separator")
			}
		})

		t.Run("should generate CSS for multiple values with compressed separator", func(t *testing.T) {
			mockValues := []any{
				&mockCSSGenerator{
					genCSSFn: func(ctx any, out *CSSOutput) {
						out.Add("test1", nil, nil)
					},
				},
				&mockCSSGenerator{
					genCSSFn: func(ctx any, out *CSSOutput) {
						out.Add("test2", nil, nil)
					},
				},
			}
			value, _ := NewValue(mockValues)
			var output []string
			outputFn := func(chunk any, fileInfo any, index any) {
				output = append(output, chunk.(string))
			}
			value.GenCSS(map[string]any{"compress": true}, &CSSOutput{Add: outputFn})
			if len(output) != 3 || output[0] != "test1" || output[1] != "," || output[2] != "test2" {
				t.Error("Expected output with compressed separator")
			}
		})

		t.Run("should handle empty/whitespace-only output", func(t *testing.T) {
			var output []string
			outputFn := func(chunk any, fileInfo any, index any) {
				output = append(output, chunk.(string))
			}
			mockValues := []any{
				&mockCSSGenerator{
					genCSSFn: func(ctx any, out *CSSOutput) {
						out.Add("", nil, nil)
					},
				},
				&mockCSSGenerator{
					genCSSFn: func(ctx any, out *CSSOutput) {
						out.Add("   ", nil, nil)
					},
				},
				&mockCSSGenerator{
					genCSSFn: func(ctx any, out *CSSOutput) {
						out.Add("test", nil, nil)
					},
				},
			}

			value, _ := NewValue(mockValues)
			value.GenCSS(map[string]any{}, &CSSOutput{Add: outputFn})

			expectedOutput := []string{"", ", ", "   ", ", ", "test"}
			if !reflect.DeepEqual(output, expectedOutput) {
				t.Error("Expected proper handling of empty/whitespace output")
			}
		})

		t.Run("should handle various context compression settings", func(t *testing.T) {
			testCases := []struct {
				context  map[string]any
				expected string
			}{
				{map[string]any{"compress": true}, ","},
				{map[string]any{"compress": false}, ", "},
				{map[string]any{}, ", "},
				{nil, ", "},
			}

			for _, tc := range testCases {
				var separator string
				outputFn := func(chunk any, fileInfo any, index any) {
					if s, ok := chunk.(string); ok && (s == "," || s == ", ") {
						separator = s
					}
				}

				mockValues := []any{
					&mockCSSGenerator{
						genCSSFn: func(ctx any, out *CSSOutput) {
							out.Add("test1", nil, nil)
						},
					},
					&mockCSSGenerator{
						genCSSFn: func(ctx any, out *CSSOutput) {
							out.Add("test2", nil, nil)
						},
					},
				}

				value, _ := NewValue(mockValues)
				value.GenCSS(tc.context, &CSSOutput{Add: outputFn})

				if separator != tc.expected {
					t.Errorf("Expected separator %q for context %v, got %q", tc.expected, tc.context, separator)
				}
			}
		})
	})
}

// Mock types for testing
type mockEvaluator struct {
	evalFn func(context any) (any, error)
}

func (m *mockEvaluator) Eval(context any) (any, error) {
	return m.evalFn(context)
}

type mockCSSGenerator struct {
	genCSSFn func(context any, output *CSSOutput)
}

func (m *mockCSSGenerator) GenCSS(context any, output *CSSOutput) {
	m.genCSSFn(context, output)
} 